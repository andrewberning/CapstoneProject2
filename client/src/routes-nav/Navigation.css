/* .Navigation {
  background-color: white;
  border-bottom: 2px solid rgba(0, 0, 0, 0.125);
}

.navbar-brand {
  margin-left: 30px;
}

.navbar-nav {
  flex-direction: row;
}
.nav-item {
  margin-right: 20px;
}
.nav-link {
  margin-right: 20px;
}
.nav-link.active {
  font-weight: bold;
  color: black;
} */