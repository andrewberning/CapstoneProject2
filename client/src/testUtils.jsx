import React from "react";
import UserContext from "./auth/UserContext";
import { vi } from 'vitest';

const demoUser = {
  username: "testuser",
  first_name: "Test",
  last_name: "User",
  email: "testuser123@test.net",
};

// Mock functions
const mockSetCurrentUser = vi.fn();
const mockLogin = vi.fn();
const mockLogout = vi.fn();
const mockSignup = vi.fn();
const mockAddToCart = vi.fn();
const mockHasAddedToCart = vi.fn();

const UserProvider = ({ 
  children,
  currentUser = demoUser,
  setCurrentUser = mockSetCurrentUser,
  login = mockLogin,
  logout = mockLogout,
  signup = mockSignup,
  addToCart = mockAddToCart,
  hasAddedToCart = mockHasAddedToCart,
  cartItems = [],
  setCartItems = () => {},
  categories = [],
  infoLoaded = true,
  totalAmount = 0,
  totalItems = 0,
}) => (
  <UserContext.Provider value={{ 
    currentUser,
    setCurrentUser,
    login,
    logout,
    signup,
    addToCart,
    hasAddedToCart,
    cartItems,
    setCartItems,
    categories,
    infoLoaded,
    totalAmount,
    totalItems,
  }}>
    {children}
  </UserContext.Provider>
);

export { UserProvider, mockSetCurrentUser, mockLogin, mockLogout, mockSignup, mockAddToCart, mockHasAddedToCart };
