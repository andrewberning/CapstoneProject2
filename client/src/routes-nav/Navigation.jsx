import { useState, useContext, useEffect } from "react";
import UserContext from "../auth/UserContext";
import SignInOffcanvas from "./SignInOffcanvas";
import LogOutOffcanvas from "./LogOutOffcanvas";
import SearchForm from "../components/SearchForm";
import Container from 'react-bootstrap/Container';
import Nav from 'react-bootstrap/Nav';
import Navbar from 'react-bootstrap/Navbar';
import Offcanvas from "react-bootstrap/Offcanvas";
import Col from 'react-bootstrap/Col';
import Row from 'react-bootstrap/Row';
import { Link } from 'react-router-dom';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faShoppingCart, faUser } from "@fortawesome/free-solid-svg-icons";
import "./Navigation.css";

export default function Navigation({ logout }) {
  const { currentUser, categories, cartItems, totalItems, setTotalItems } = useContext(UserContext);
  const [showOffcanvas, setShowOffcanvas] = useState(false);
  const [showSignIn, setShowSignIn] = useState(false);
  const [showUserMenu, setShowUserMenu] = useState(false);
  // const [cartCount, setCartCount] = useState(0);

  const handleOffcanvasClose = () => setShowOffcanvas(false);
  const handleOffcanvasShow = () => setShowOffcanvas(true);

  const handleSignInClose = () => setShowSignIn(false);
  const handleSignInShow = () => setShowSignIn(true);

  const handleUserMenuClose = () => setShowUserMenu(false);
  const handleUserMenuShow = () => setShowUserMenu(true);

  const handleLogout = () => {
    logout();
    handleUserMenuClose();
  }

  useEffect(() => {
    const itemCount = cartItems.reduce((total, item) => total + item.quantity, 0);
    setTotalItems(itemCount);
  }, []);

  return (
    <Navbar expand={false} className="bg-body-tertiary">
    <Container fluid>
      <Row className="w-100 align-items-center">
        <Col xs="auto">
          <Navbar.Brand href="/">Shoply</Navbar.Brand>
        </Col>
        <Col xs="auto">
          <Navbar.Toggle aria-controls="offcanvasNavbar" onClick={handleOffcanvasShow} />
        </Col>
        <Col className="flex-grow-1">
          <SearchForm />
        </Col>
        <Col xs="auto">
          <Nav className="ms-auto">
            {currentUser 
             ? <Nav.Link href="" onClick={handleUserMenuShow}>
               <FontAwesomeIcon icon={faUser} /> {currentUser.username}
               </Nav.Link>
             : <Nav.Link href="" onClick={handleSignInShow}>Sign In</Nav.Link>
            }
          </Nav>
        </Col>
        <Col xs="auto">
          <Nav className="ms-auto">
            <Link to="/cart" className="nav-link position-relative">
            <FontAwesomeIcon icon={faShoppingCart} />
            {totalItems > 0 && (
              <span className="position-absolute top-0 start-100 translate-middle badge rounded-pill bg-danger">
                {totalItems}
              </span>
            )}
            </Link>
          </Nav>
        </Col>
      </Row>
    </Container>
    <Navbar.Offcanvas
      id="offcanvasNavbar-expand"
      show={showOffcanvas}
      onHide={handleOffcanvasClose}
      aria-labelledby="offcanvasNavbarLabel"
      placement="start"
    >
      <Offcanvas.Header closeButton>
        <Offcanvas.Title id="offcanvasNavbarLabel-expand">
          <Link to={"/"} className="nav-link" onClick={handleOffcanvasClose}>
            Shoply
          </Link>
        </Offcanvas.Title>
      </Offcanvas.Header>
      <Offcanvas.Body>
        <Nav className="me-auto flex-column">
          {/* Link to the categories list page */}
          <Nav.Link as="div">
            <Link to="/categories" className="nav-link" onClick={handleOffcanvasClose}>
              Categories
            </Link>
          </Nav.Link>

          {/* List all categories as links */}
          {categories.map((category) => (
            <Nav.Link as="div" key={category.id}>
              <Link 
                to={`/categories/${category.name}`} 
                className="nav-link ms-3"
                onClick={handleOffcanvasClose}
              >
                {category.name}
              </Link>
            </Nav.Link>
          ))}
        </Nav>
      </Offcanvas.Body>
    </Navbar.Offcanvas>

    {/* Sign In Offcanvas */}
    <SignInOffcanvas show={showSignIn} handleClose={handleSignInClose} />

    <LogOutOffcanvas show={showUserMenu} handleClose={handleUserMenuClose} logout={handleLogout} />
  </Navbar>
  );
}