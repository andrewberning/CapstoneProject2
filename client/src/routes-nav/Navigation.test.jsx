import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import Navigation from "./Navigation";
import { UserProvider } from "../testUtils";


describe('Navigation component', () => {
  it("renders without crashing", function () {
    render(
      <MemoryRouter>
        <UserProvider >
          <Navigation />
        </UserProvider>
      </MemoryRouter>
    );
  });

  it("matches snapshot", function () {
    const { asFragment } = render(
      <MemoryRouter>
        <UserProvider >
          <Navigation />
        </UserProvider>
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot when logged out", function () {
    const { asFragment } = render(
      <MemoryRouter>
        <UserProvider >
          <Navigation />
        </UserProvider>
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});