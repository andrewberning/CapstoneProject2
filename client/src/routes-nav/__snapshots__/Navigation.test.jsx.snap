// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Navigation component > matches snapshot 1`] = `
<DocumentFragment>
  <nav
    class="bg-body-tertiary navbar navbar-light"
  >
    <div
      class="container-fluid"
    >
      <div
        class="w-100 align-items-center row"
      >
        <div
          class="col-auto"
        >
          <a
            class="navbar-brand"
            href="/"
          >
            Shoply
          </a>
        </div>
        <div
          class="col-auto"
        >
          <button
            aria-controls="offcanvasNavbar"
            aria-label="Toggle navigation"
            class="navbar-toggler collapsed"
            type="button"
          >
            <span
              class="navbar-toggler-icon"
            />
          </button>
        </div>
        <div
          class="flex-grow-1 col"
        >
          <div
            class="SearchForm"
          >
            <form
              class="search-form d-flex align-items-center"
              role="search"
            >
              <div
                class="input-group"
              >
                <input
                  aria-label="Search"
                  class="form-control"
                  name="searchTerm"
                  placeholder="Search..."
                  type="text"
                  value=""
                />
                <button
                  class="input-group-text"
                >
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-magnifying-glass "
                    data-icon="magnifying-glass"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M416 208c0 45.9-14.9 88.3-40 122.7L502.6 457.4c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L330.7 376c-34.4 25.2-76.8 40-122.7 40C93.1 416 0 322.9 0 208S93.1 0 208 0S416 93.1 416 208zM208 352a144 144 0 1 0 0-288 144 144 0 1 0 0 288z"
                      fill="currentColor"
                    />
                  </svg>
                </button>
              </div>
            </form>
          </div>
        </div>
        <div
          class="col-auto"
        >
          <div
            class="ms-auto navbar-nav"
          >
            <a
              class="nav-link"
              href="#"
              role="button"
              tabindex="0"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-user "
                data-icon="user"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M224 256A128 128 0 1 0 224 0a128 128 0 1 0 0 256zm-45.7 48C79.8 304 0 383.8 0 482.3C0 498.7 13.3 512 29.7 512H418.3c16.4 0 29.7-13.3 29.7-29.7C448 383.8 368.2 304 269.7 304H178.3z"
                  fill="currentColor"
                />
              </svg>
               testuser
            </a>
          </div>
        </div>
        <div
          class="col-auto"
        >
          <div
            class="ms-auto navbar-nav"
          >
            <a
              class="nav-link"
              data-rr-ui-event-key="/cart"
              href="/cart"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-cart-shopping "
                data-icon="cart-shopping"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 576 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 24C0 10.7 10.7 0 24 0H69.5c22 0 41.5 12.8 50.6 32h411c26.3 0 45.5 25 38.6 50.4l-41 152.3c-8.5 31.4-37 53.3-69.5 53.3H170.7l5.4 28.5c2.2 11.3 12.1 19.5 23.6 19.5H488c13.3 0 24 10.7 24 24s-10.7 24-24 24H199.7c-34.6 0-64.3-24.6-70.7-58.5L77.4 54.5c-.7-3.8-4-6.5-7.9-6.5H24C10.7 48 0 37.3 0 24zM128 464a48 48 0 1 1 96 0 48 48 0 1 1 -96 0zm336-48a48 48 0 1 1 0 96 48 48 0 1 1 0-96z"
                  fill="currentColor"
                />
              </svg>
            </a>
          </div>
        </div>
      </div>
    </div>
    <div
      aria-labelledby="offcanvasNavbarLabel"
      class="offcanvas offcanvas-start"
      id="offcanvasNavbar-expand"
    >
      <div
        class="offcanvas-header"
      >
        <div
          class="offcanvas-title h5"
          id="offcanvasNavbarLabel-expand"
        >
          Shoply
        </div>
        <button
          aria-label="Close"
          class="btn-close"
          type="button"
        />
      </div>
      <div
        class="offcanvas-body"
      >
        <div
          class="me-auto navbar-nav"
        >
          <div
            class="nav-item dropdown"
          >
            <a
              aria-expanded="false"
              class="dropdown-toggle nav-link"
              href="#"
              id="offcanvasNavbarDropdown-expand"
              role="button"
              tabindex="0"
            >
              Categories
            </a>
          </div>
        </div>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;

exports[`Navigation component > matches snapshot when logged out 1`] = `
<DocumentFragment>
  <nav
    class="bg-body-tertiary navbar navbar-light"
  >
    <div
      class="container-fluid"
    >
      <div
        class="w-100 align-items-center row"
      >
        <div
          class="col-auto"
        >
          <a
            class="navbar-brand"
            href="/"
          >
            Shoply
          </a>
        </div>
        <div
          class="col-auto"
        >
          <button
            aria-controls="offcanvasNavbar"
            aria-label="Toggle navigation"
            class="navbar-toggler collapsed"
            type="button"
          >
            <span
              class="navbar-toggler-icon"
            />
          </button>
        </div>
        <div
          class="flex-grow-1 col"
        >
          <div
            class="SearchForm"
          >
            <form
              class="search-form d-flex align-items-center"
              role="search"
            >
              <div
                class="input-group"
              >
                <input
                  aria-label="Search"
                  class="form-control"
                  name="searchTerm"
                  placeholder="Search..."
                  type="text"
                  value=""
                />
                <button
                  class="input-group-text"
                >
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-magnifying-glass "
                    data-icon="magnifying-glass"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M416 208c0 45.9-14.9 88.3-40 122.7L502.6 457.4c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L330.7 376c-34.4 25.2-76.8 40-122.7 40C93.1 416 0 322.9 0 208S93.1 0 208 0S416 93.1 416 208zM208 352a144 144 0 1 0 0-288 144 144 0 1 0 0 288z"
                      fill="currentColor"
                    />
                  </svg>
                </button>
              </div>
            </form>
          </div>
        </div>
        <div
          class="col-auto"
        >
          <div
            class="ms-auto navbar-nav"
          >
            <a
              class="nav-link"
              href="#"
              role="button"
              tabindex="0"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-user "
                data-icon="user"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M224 256A128 128 0 1 0 224 0a128 128 0 1 0 0 256zm-45.7 48C79.8 304 0 383.8 0 482.3C0 498.7 13.3 512 29.7 512H418.3c16.4 0 29.7-13.3 29.7-29.7C448 383.8 368.2 304 269.7 304H178.3z"
                  fill="currentColor"
                />
              </svg>
               testuser
            </a>
          </div>
        </div>
        <div
          class="col-auto"
        >
          <div
            class="ms-auto navbar-nav"
          >
            <a
              class="nav-link"
              data-rr-ui-event-key="/cart"
              href="/cart"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-cart-shopping "
                data-icon="cart-shopping"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 576 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 24C0 10.7 10.7 0 24 0H69.5c22 0 41.5 12.8 50.6 32h411c26.3 0 45.5 25 38.6 50.4l-41 152.3c-8.5 31.4-37 53.3-69.5 53.3H170.7l5.4 28.5c2.2 11.3 12.1 19.5 23.6 19.5H488c13.3 0 24 10.7 24 24s-10.7 24-24 24H199.7c-34.6 0-64.3-24.6-70.7-58.5L77.4 54.5c-.7-3.8-4-6.5-7.9-6.5H24C10.7 48 0 37.3 0 24zM128 464a48 48 0 1 1 96 0 48 48 0 1 1 -96 0zm336-48a48 48 0 1 1 0 96 48 48 0 1 1 0-96z"
                  fill="currentColor"
                />
              </svg>
            </a>
          </div>
        </div>
      </div>
    </div>
    <div
      aria-labelledby="offcanvasNavbarLabel"
      class="offcanvas offcanvas-start"
      id="offcanvasNavbar-expand"
    >
      <div
        class="offcanvas-header"
      >
        <div
          class="offcanvas-title h5"
          id="offcanvasNavbarLabel-expand"
        >
          Shoply
        </div>
        <button
          aria-label="Close"
          class="btn-close"
          type="button"
        />
      </div>
      <div
        class="offcanvas-body"
      >
        <div
          class="me-auto navbar-nav"
        >
          <div
            class="nav-item dropdown"
          >
            <a
              aria-expanded="false"
              class="dropdown-toggle nav-link"
              href="#"
              id="offcanvasNavbarDropdown-expand"
              role="button"
              tabindex="0"
            >
              Categories
            </a>
          </div>
        </div>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;
